### PR TITLE
Setup repository

### PR DESCRIPTION
**Pull request by ptrdom**

### CODE REVIEW

**Comment by ptrdom** [comment]
TODO:
- [x]  Setup CI tests
- [x]  Setup CI publishing
- [x]  Setup documentation

File: build.sbt [hunk 1]
```diff
@@ -1,53 +1,94 @@
...
+
+val akkaActorVersion = "2.6.17"
+val akkaHttpVersion = "10.2.6"
+
+inThisBuild(
+  List(
+    versionPolicyIntention := Compatibility.BinaryAndSourceCompatible,
+    organization := "org.endpoints4s",
+    sonatypeProjectHosting := Some(
+      GitHubHosting("endpoints4s", "play", "julien@richard-foy.fr")
```
Comment (julienrf) [inline build.sbt]:
```suggestion
      GitHubHosting("endpoints4s", "akka-http", "julien@richard-foy.fr")
```

:wink:

**Comment by ptrdom** [comment]
Closing and reopening the PR is a good trick to get the workflows trigger for the first time 😄

**Comment by ptrdom** [comment]
@julienrf Could you take a look at Scala 3 setup? I think we might have an issue because we do not publish Scala 3 versions of `algebra-testkit` and `algebra-circe-testkit`.

File: build.sbt [hunk 2]
```diff
@@ -1,53 +1,94 @@
-import EndpointsSettings._
+import xerial.sbt.Sonatype.GitHubHosting
 
-val `algebra-jvm` = LocalProject("algebraJVM")
-val `algebra-testkit-jvm` = LocalProject("algebra-testkitJVM")
-val `algebra-circe-testkit-jvm` = LocalProject("algebra-circe-testkitJVM")
-val `json-schema-generic-jvm` = LocalProject("json-schema-genericJVM")
-val `openapi-jvm` = LocalProject("openapiJVM")
+val algebra = "org.endpoints4s" %% "algebra" % "1.10.0"
+val algebraTestkit = "org.endpoints4s" %% "algebra-testkit" % "4.1.0"
```
Comment (julienrf) [inline build.sbt:4]: If that does not work with Scala 3, it should be possible to use the `Cross.for3Use2_13` trick, no?

File: build.sbt [hunk 3]
```diff
@@ -1,53 +1,94 @@
...
     .in(file("server"))
     .settings(
-      publishSettings,
-      `scala 2.12 to dotty`,
       name := "akka-http-server",
       publish / skip := scalaBinaryVersion.value.startsWith("3"),
       libraryDependencies ++= Seq(
+        algebra,
+        openapi,
+        algebraTestkit % Test,
```
Comment (julienrf) [inline build.sbt]: If that does not work with Scala 3, it should be possible to use the `CrossVersion.for3Use2_13` trick, no?

```suggestion
        (algebraTestkit % Test).cross(CrossVersion.for3Use2_13),
```

Comment (ptrdom) [reply build.sbt]: That does not work - the Scala 3 `compile` starts throwing various Scala 3 errors with `CrossVersion.for3Use2_13`.

Comment (ptrdom) [reply build.sbt]: Please check latest CI run for errors.

**Comment by julienrf** [comment]
Ah, I see, it must be because of this exclusion rule here:

https://github.com/endpoints4s/akka-http/blob/633a3b646011f115c4dbe5d532c25bbdf0b186f2/build.sbt#L26

It was added because Scala 3 modules were (transitively) depending on both `scala-collection-compat_2.13` and `scala-collection-compat_3`. It seems the Scala 3 artifact is not anymore in the classpath now? What happens if you remove that exclusion rule in both modules `akka-http-client` and `akka-http-server`?

**Comment by ptrdom** [comment]
Now a different set of errors is thrown.

**Comment by julienrf** [comment]
Thank you so much for your patience @ptrdom! This is a complicated situation. Some of the transitive dependencies are for Scala 2, and in general this is not a problem except when they define macros that we want to use. I believe this is what happens here: the generic JSON schema derivation macros are the ones for Scala 2, and the Scalatest failures seem to be related to a problem to synthesize an implicit value by a macro.

I am really sorry that you have to handle such a complex setup… I believe the correct solution would be to publish the testkit modules for Scala 3 in the core project, and then depend on that here. What do you think? Let me see if I can backpublish them (not sure I will manage to do that before a couple of days).

**Comment by julienrf** [comment]
Ah, I see that `algebra-testkit` has dependencies on Scala 2.13 modules… This is why we could not publish for Scala 3 in the first place. Then I think for now the solution should be to “play” with the dependencies to exclude the wrong ones and include the right ones… Please let me know if you could do that, otherwise I may have a look next week.

**Comment by ptrdom** [comment]
~~No worries! I think the solution of publishing testkit modules for Scala 3 is the right one. Please let me know when the dependencies are published, in the meanwhile I will hopefully have time to look into remaining items - publishing and documentation.~~

Aha, I see the situation changed again. So if a module contains `cross(CrossVersion.for3Use2_13)` dependencies then it cannot be published?

**Comment by ptrdom** [comment]
After reading https://www.scala-lang.org/blog/2021/04/08/scala-3-in-sbt.html#using-scala-213-libraries-in-scala-3 I wonder what issues publishing `algebra-testkit` modules for Scala 3 as they are now would cause? Maybe our case falls under the edge case where we "know what we are doing"? 😄

**Comment by ptrdom** [comment]
The only expected users of `algebra-testkit` modules are other endpoints4s modules, specifically the `test` scopes of them, so I do not see how a problematic version mix-up can happen there.

**Comment by ptrdom** [comment]
Or we can do the same as what is done in `play` interpreter - we only compile for Scala 3, but do not publish nor test.

**Comment by julienrf** [comment]
> Or we can do the same as what is done in `play` interpreter - we only compile for Scala 3, but do not publish nor test.

I agree, that's probably the best idea!

**Comment by ptrdom** [comment]
Works!

File: documentation/src/main/paradox/index.md [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+### `endpoints4s.akkahttp.client.ChunkedEntities`
+
+The `ChunkedEntities` interpreter fixes the `Chunks[A]` type to `akka.stream.scaladsl.Source[A, _]`:
+
+@@snip [ChunkedEntities.scala](/client/src/main/scala/endpoints4s/akkahttp/client/ChunkedEntities.scala) { #stream-type }
+
+This means that, given the following endpoint definition:
+
+@@snip [ChunkedEntitiesDocs.scala](/algebras/algebra/src/test/scala/endpoints4s/algebra/ChunkedEntitiesDocs.scala) { #streamed-endpoint }
```
Comment (ptrdom) [inline documentation/src/main/paradox/index.md]: @julienrf Any tips on how to deal with these links to original repo?

File: documentation/src/main/paradox/index.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+Alternatively, there is also a method `implementedByAsync` that takes an implementing function
+returning a `Future[B]`.
+
+### `endpoints4s.akkahttp.server.ChunkedEntities`
+
+The `ChunkedEntities` interpreter fixes the `Chunks[A]` type to `akka.stream.scaladsl.Source[A, _]`.
+
+For instance, given the following chunked endpoint definition:
+
+@@snip [ChunkedEntitiesDocs.scala](/algebras/algebra/src/test/scala/endpoints4s/algebra/ChunkedEntitiesDocs.scala) { #streamed-endpoint }
```
Comment (ptrdom) [inline documentation/src/main/paradox/index.md]: Same as https://github.com/endpoints4s/akka-http/pull/1/files#r1398495601.

File: documentation/src/main/paradox/index.md (hunk 4 above)
Comment (julienrf) [reply documentation/src/main/paradox/index.md]: What I did in the `play` repo is to copy-paste the endpoint definition so that the example is fully self-contained in the repo.

https://github.com/endpoints4s/play/blob/v5.0.0/client/src/test/scala/endpoints4s/play/client/EndpointsDocs.scala#L9-L10

We could do the same here.

**Comment by ptrdom** [comment]
@julienrf I think the PR is ready. One thing I cannot fully check is the release process, I am guessing we will iron any issues out once the code is in `main`.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    name: Test (Scala 2.12)
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-java@v3
+        with:
+          distribution: temurin
+          java-version: 11
+          cache: sbt
+      - run: npm install jsdom@12.2.0
```
Comment (julienrf) [inline .github/workflows/ci.yml]: Do we need `jsdom`?

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+          distribution: temurin
+          java-version: 11
+          cache: sbt
+      - run: sbt versionCheck
+        env:
+          PGP_PASSPHRASE: ${{ secrets.PGP_PASSPHRASE }}
+          PGP_SECRET: ${{ secrets.PGP_SECRET }}
+          SONATYPE_PASSWORD: ${{ secrets.SONATYPE_PASSWORD }}
+          SONATYPE_USERNAME: ${{ secrets.SONATYPE_USERNAME }}
+      - run: sbt ci-release
```
Comment (julienrf) [inline .github/workflows/release.yml]: It seems the `env` applies to the `versionCheck` step, not the `ci-release` one.

File: build.sbt [hunk 8]
```diff
@@ -1,67 +1,162 @@
...
+          (ThisBuild / sonatypeProjectHosting).value.get.scmInfo.browseUrl.toURI
+        theme
+          .withRepository(repository)
+          .withSocial(repository)
+          .withCustomStylesheet("snippets.css")
+      },
+      paradoxProperties ++= Map(
+        "version" -> version.value,
+        "scaladoc.base_url" -> s".../${(packageDoc / siteSubdirName).value}",
+        "github.base_url" -> s"${homepage.value.get}/blob/v${version.value}"
```
Comment (julienrf) [inline build.sbt:135]: Don’t we need to also pass the `akka-version` variable?

**Review by julienrf — COMMENTED** [review]
Thank you very much for all the work! There are a couple of minor mistakes.

File: .github/workflows/release.yml (hunk 7 above)
Comment (ptrdom) [reply .github/workflows/release.yml]: `Play` release is broken too then.

File: .github/workflows/ci.yml (hunk 6 above)
Comment (ptrdom) [reply .github/workflows/ci.yml]: Likely not, I think only `js` intepreters need it.

File: build.sbt (hunk 8 above)
Comment (ptrdom) [reply build.sbt:135]: I checked - seems like it is needed.

**Review by julienrf — APPROVED** [review]
Awesome, thank you very much @ptrdom!

**Comment by julienrf** [comment]
Let’s merge it! If there is any remaining issue we will fix it afterwards…
We should not cut a release but instead we should manually deploy the website of the last published version.

File: .github/workflows/release.yml (hunk 7 above)
Comment (julienrf) [reply .github/workflows/release.yml]: Thank you, I edited the file there too!

**Comment by ptrdom** [comment]
Published the site, seems to work fine.

**Comment by ptrdom** [comment]
Guessing that the versions in docs and links to artifacts will fix themselves after first release from this branch. Maybe even a tag would be enough.
